pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -570,14 +570,14 @@ void refreshSegment(String segment) {
     }
 
     InstanceSelector.SelectionResult calculateRouting(BrokerRequest brokerRequest) {
-      List<String> selectedSegments = _segmentSelector.select(brokerRequest);
+      Set<String> selectedSegments = _segmentSelector.select(brokerRequest);
       if (!selectedSegments.isEmpty()) {
         for (SegmentPruner segmentPruner : _segmentPruners) {
           selectedSegments = segmentPruner.prune(brokerRequest, selectedSegments);
         }
       }
       if (!selectedSegments.isEmpty()) {
-        return _instanceSelector.select(brokerRequest, selectedSegments);
+        return _instanceSelector.select(brokerRequest, new ArrayList<>(selectedSegments));
```

Comment:
Why did you change the interface to return `Set` for `SegmentSelector` if you will anyway convert it back to the array list?

Reply:
Current `BalancedInstanceSelector` logic relies on the order of the selected segments, maybe we can change that logic and change this api to be `Set`.